README.md
## Composer — Reactive Android Instrumentation Test Runner.

Composer is a modern reactive replacement for [square/spoon][spoon] with following feature set:

* Parallel test execution on multiple emulators/devices with [test sharding][test sharding] support.
* Logcat output capturing per test and for whole test run as well.
* Screenshots and files pulling for each test reactively (with support for [square/spoon][spoon] folder structure).
* JUnit4 report generation. 

![Demo](demo/composer.gif)

### Fork additions

This repo is a fork of [gojuno/commander](https://github.com/gojuno/commander). Additions I (Nilzor) have made:

- Added Windows support by improving command line interfacing and folder name scheme
- Added command line parameter `--device-aliases` (see below)

### Table of Contents

- [Why we've decided to replace square/spoon](#why-weve-decided-to-replace-squarespoon)
- [HTML Report](#html-report)
- [Usage](#usage)
- [Download](#download)
- [3rd-party Composer Gradle Plugin](#3rd-party-composer-gradle-plugin)
- [Swarmer](#swarmer)
- [How to build](#how-to-build)
- [License](#license)

### Why we've decided to replace [square/spoon][spoon]
 
**Problem 1:** Our UI tests are stable, but we saw a lot of UI tests build failures. About ~50% of our CI builds were failing. All such failures of UI tests came from Spoon not being able to run tests on one or more emulators (device is red in the report and error message is `…work/emulator-5554/result.json (No such file or directory)`, basically it timed out on installing the apk on a device, increasing adb timeout did not help, all emulators responded to adb commands and mouse/keyboard interactions, we suppose problem is in in ddmlib used by Spoon.

**Solution:** Composer does not use ddmlib and talks to emulators/devices by invoking `adb` binary.  

**Problem 2:** Pretty often when test run finished, Spoon freezed on moving screenshots from one of the emulators/devices. Again, we blame ddmlib used in Spoon for that.

**Solution:** Composer invokes `adb` binary to pull files from emulators/devices, we haven't seen problems with that in more than 700 builds on CI.

**Problem 3:** Spoon pulled screenshots/files *after* finish of the whole test run on a device which slows down builds: `test_run_time + pull_files_time`.

**Solution:** Composer pulls screenshots/files *reactively* after each test which basically leads to: `~test_run_time`.

**Problem 4:** If test sharding is enabled (which we do all the time), Spoon HTML report is very hard to look at, especially if you want to find some particular test(s) and it's not failed. You have to either hover mouse over each test to find out its name or go into html/xml source and find on which emulator/device test was sharded in order to click on correct device and then find test by CMD+F on the page.

**Solution:** HTML report we've built designed with usability and performance in mind.
  
**Problem 5:** Html report can be very slow to load if you have lots of screenshots (which we do) since it displays all the screenshots of tests that were run on a particular device on a single page — it can take up to minutes to finish while you effectively unable to scroll page since scroll is jumping up and down each time new screenshot loaded.

**Solution:** HTML report that we've built does not display screenshots on index and suite pages, screenshots are displayed only on the test page → fast page load.

>With Composer we were able to make UI tests required part of CI for Pull Requests.
>It's fast, reliable and uses RxJava which means that it's relatively easy to add more features combining complex async transformations. 

### HTML Report

Our Frontend Team [helped us](https://github.com/gojuno/composer/issues/11) build HTML Report for the Composer. 

>It's fast, small and designed in collaboration with our QAs and Developers who actually use it on daily basis to make it easy to use.

Here are few screenshots:

[<img src="demo/screenshot1.png" alt="Suite Page" style="height: 300px;"/>](demo/screenshot1.png) [<img src="demo/screenshot2.png" alt="Test Page" style="height: 300px;"/>](demo/screenshot2.png)[<img src="demo/screenshot3.png" alt="Test Page" style="height: 300px;"/>](demo/screenshot3.png)

## Usage

Composer shipped as jar, to run it you need JVM 1.8+: `java -jar composer-latest-version.jar options`. 

#### Supported options

##### Required

* `--apk`
  * Either relative or absolute path to application apk that needs to be tested.
  * Example: `--apk myapp.apk`
* `--test-apk`
  * Either relative or absolute path to apk with tests.
  * Example: `--test-apk myapp-androidTest.apk`

##### Optional

* `--help, -help, help, -h`
  * Print help and exit.
* `--test-runner`
  * Fully qualified name of test runner class you're using. 
  * Default: automatically parsed from `--test-apk`'s `AndroidManifest`. 
  * Example: `--test-runner com.example.TestRunner`
* `--shard`
  * Either `true` or `false` to enable/disable [test sharding][test sharding] which statically shards tests between available devices/emulators. 
  * Default: `true`.
  * Example: `--shard false`
* `--output-directory`
  * Either relative or absolute path to directory for output: reports, files from devices and so on. 
  * Default: `composer-output` in current working directory.
  * Example: `--output-directory artifacts/composer-output`
* `--instrumentation-arguments`
  * Key-value pairs to pass to Instrumentation Runner.
  * Default: empty.
  * Example: `--instrumentation-arguments myKey1 myValue1 myKey2 myValue2`.
* `--verbose-output`
  * Either `true` or `false` to enable/disable verbose output for Composer. 
  * Default: `false`.
  * Example: `--verbose-output true`
* `--keep-output-on-exit`
  * Either `true` or `false` to keep/clean temporary output files used by Composer on exit. 
  * Default: `false`.
  * Composer uses files to pipe output of external commands like `adb`, keeping them might be useful for debugging issues. 
  * Example: `--keep-output-on-exit true`
* `--devices`
  * Connected devices/emulators that will be used to run tests against.
  * Default: empty, tests will run on all connected devices/emulators.
  * Specifying both `--devices` and `--device-pattern` will result in an error.
  * Example: `--devices emulator-5554 emulator-5556`
* `--device-pattern`
  * Connected devices/emulators that will be used to run tests against.
  * Default: empty, tests will run on all connected devices/emulators.
  * Specifying both `--device-pattern` and `--devices` will result in an error.
  * Example: `--device-pattern "emulator.+"`
* `--device-aliases`
  * List of comma separated key-value pairs of device ID aliases. Will replace the Android Device IDs in the suite list (but not in report details). 
  * Example: `--device-aliases "123QR=SamsungS10,345AF=SamsungS20"`
* `--install-timeout`
  * APK installation timeout in seconds. 
  * Default: `120` seconds (2 minutes). 
  * Applicable to both test APK and APK under test. 
  * Example: `--install-timeout 20`
* `--fail-if-no-tests`
  * Either `true` or `false` to enable/disable error on empty test suite.
  * Default: `true`.
  * `False` may be applicable when you run tests conditionally(via annotation/package filters) and empty suite is a valid outcome.
  * Example: `--fail-if-no-tests false`
* `--with-orchestrator`
  * Either `true` or `false` to enable/disable running tests via AndroidX Test Orchestrator.
  * Default: `false`.
  * When enabled - minimizes shared state and isolates test crashes.
  * Requires test orchestrator & test services APKs to be installed on device before executing.
  * More info: https://developer.android.com/training/testing/junit-runner#using-android-test-orchestrator
  * Example: `--with-orchestrator true`
* `--extra-apks`
  * Apks to be installed for utilities. What you would typically declare in gradle as `androidTestUtil` 
  * Default: empty, only apk and test apk would be installed.
  * Works great with Orchestrator to install orchestrator & test services APKs.
  * Example: `--extra-apks path/to/apk/first.apk path/to/apk/second.apk`
  
##### Example

Simplest :
```console
java -jar composer-latest-version.jar \
--apk app/build/outputs/apk/example-debug.apk \
--test-apk app/build/outputs/apk/example-debug-androidTest.apk
```

With arguments :
```console
java -jar composer-latest-version.jar \
--apk app/build/outputs/apk/example-debug.apk \
--test-apk app/build/outputs/apk/example-debug-androidTest.apk \
--test-runner com.example.test.ExampleTestRunner \
--output-directory artifacts/composer-output \
--instrumentation-arguments key1 value1 key2 value2 \
--verbose-output false \
--keep-output-on-exit false \
--with-orchestrator false
```

### Download

- See [Releases](https://github.com/nilzor/composer/releases)
 
### 3rd-party Composer Gradle Plugin

[@trevjonez](https://github.com/trevjonez) [built](https://github.com/gojuno/composer/issues/77) 🎉 [Gradle Plugin for Composer](https://github.com/trevjonez/composer-gradle-plugin) which allows you to configure and run Composer with Gradle.

### Swarmer

Composer works great in combination with [Swarmer][swarmer] — another tool we've built at Juno. 

[Swarmer][swarmer] can create and start multiple emulators in parallel. In our [CI Pipeline][ci pipeline] we start emulators with Swarmer and then Composer runs tests on them.

### How to build

#### All-in-one script (used in Travis build)

Dependencies: `docker` and `bash`.

```console
ci/build.sh
```

#### Build Composer

Environment variable `ANDROID_HOME` must be set.

```console
./gradlew build
```

#### Build HTML report module

Dependencies: `npm` and `nodejs`.

```console
cd html-report
npm install
npm build
```

## License

```
Copyright 2017 Juno, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

[spoon]: https://github.com/square/spoon
[test sharding]: https://developer.android.com/training/testing/junit-runner.html#sharding-tests
[swarmer]: https://github.com/gojuno/swarmer
[ci pipeline]: https://github.com/gojuno/engineering/tree/master/articles/ci_pipeline_and_custom_tools_of_android_projects


CHANGELOG.md
# See [Releases](https://github.com/nilzor/composer/releases)


composer/src/main/kotlin/com/gojuno/composer/Args.kt
package com.gojuno.composer

import com.beust.jcommander.IStringConverter
import com.beust.jcommander.JCommander
import com.beust.jcommander.Parameter
import java.util.concurrent.TimeUnit

data class Args(
        @Parameter(
                names = arrayOf("--apk"),
                required = true,
                description = "Path to application apk that needs to be tested.",
                order = 0
        )
        var appApkPath: String = "",

        @Parameter(
                names = arrayOf("--test-apk"),
                required = true,
                description = "Path to apk with tests.",
                order = 1
        )
        var testApkPath: String = "",

        @Parameter(
                names = arrayOf("--test-runner"),
                required = false,
                description = "Fully qualified name of test runner class you're using. Will be parsed from test APK if not specified excplicitly.",
                order = 2
        )
        var testRunner: String = "",

        @Parameter(
                names = arrayOf("--shard"),
                required = false,
                arity = 1,
                description = "Either `true` or `false` to enable/disable test sharding which runs tests in parallel on available devices/emulators. `true` by default.",
                order = 3
        )
        var shard: Boolean = true,

        @Parameter(
                names = arrayOf("--output-directory"),
                required = false,
                description = "Either relative or absolute path to directory for output: reports, files from devices and so on. `composer-output` by default.",
                order = 4
        )
        var outputDirectory: String = "composer-output",

        @Parameter(
                names = arrayOf("--instrumentation-arguments"),
                required = false,
                variableArity = true,
                description = "Key-value pairs to pass to Instrumentation Runner. Usage example: `--instrumentation-arguments myKey1 myValue1 myKey2 myValue2`.",
                listConverter = InstrumentationArgumentsConverter::class,
                order = 5
        )
        var instrumentationArguments: List<String> = listOf(),

        @Parameter(
                names = arrayOf("--verbose-output"),
                required = false,
                arity = 1,
                description = "Either `true` or `false` to enable/disable verbose output for Composer. `false` by default.",
                order = 6
        )
        var verboseOutput: Boolean = false,

        @Parameter(
                names = arrayOf("--keep-output-on-exit"),
                required = false,
                description = "Either `true` or `false` to keep/clean output on exit. `false` by default.",
                order = 7
        )
        var keepOutputOnExit: Boolean = false,

        @Parameter(
                names = arrayOf("--devices"),
                required = false,
                variableArity = true,
                description = "Connected devices/emulators that will be used to run tests against. If not passed — tests will run on all connected devices/emulators. Specifying both `--devices` and `--device-pattern` will result in an error. Usage example: `--devices emulator-5554 emulator-5556`.",
                order = 8
        )
        var devices: List<String> = emptyList(),

        @Parameter(
                names = arrayOf("--device-pattern"),
                required = false,
                description = "Connected devices/emulators that will be used to run tests against. If not passed — tests will run on all connected devices/emulators. Specifying both `--device-pattern` and `--devices` will result in an error. Usage example: `--device-pattern \"somePatterns\"`.",
                order = 14
        )
        var devicePattern: String = "",

        @Parameter(
                names = arrayOf("--install-timeout"),
                required = false,
                description = "APK installation timeout in seconds. If not passed defaults to 120 seconds (2 minutes). Applicable to both test APK and APK under test.",
                order = 10
        )
        var installTimeoutSeconds: Int = TimeUnit.MINUTES.toSeconds(2).toInt(),

        @Parameter(
                names = arrayOf("--fail-if-no-tests"),
                required = false,
                arity = 1,
                description = "Either `true` or `false` to enable/disable error on empty test suite. True by default.",
                order = 11
        )
        var failIfNoTests: Boolean = true,

        @Parameter(
                names = arrayOf("--with-orchestrator"),
                required = false,
                arity = 1,
                description = "Either `true` or `false` to enable/disable running tests via AndroidX Test Orchestrator. False by default.",
                order = 12
        )
        var runWithOrchestrator: Boolean = false,

        @Parameter(
            names = arrayOf("--extra-apks"),
            required = false,
            variableArity = true,
            description = "Extra APKs you would usually put on androidTestUtil",
            order = 13
        )
        var extraApks: List<String> = emptyList(),

        @Parameter(
                names = arrayOf("--device-aliases"),
                required = false,
                variableArity = true,
                description = "Map of aliases for device ID's. E.g. '1454324KZZ32=Alpha,765FF2125=Bravo'",
                order = 8
        )
        var deviceAliases: List<String> = emptyList()
) {
        /** Converts the list of key-value pairs stored as String, to a Map */
        val deviceAliasMap: Map<String, String> by lazy {
                val map = mutableMapOf<String, String>()
                for (deviceAlias in deviceAliases) {
                        val keyVal = deviceAlias.split("=")
                        if (keyVal.size >= 2) {
                                map.put(keyVal[0], keyVal[1])
                        }
                }
                map
        }
}

// No way to share array both for runtime and annotation without reflection.
private val PARAMETER_HELP_NAMES = setOf("--help", "-help", "help", "-h")

private fun validateArguments(args: Args) {
    if (!args.devicePattern.isEmpty() && !args.devices.isEmpty()) {
        throw IllegalArgumentException("Specifying both --devices and --device-pattern is prohibited.")
    }
}

fun parseArgs(rawArgs: Array<String>) = Args().also { args ->
    if (PARAMETER_HELP_NAMES.any { rawArgs.contains(it) }) {
        JCommander(args).usage()
        exit(Exit.Ok)
    }

    JCommander.newBuilder()
            .addObject(args)
            .build()
            .parse(*rawArgs)
    validateArguments(args)
}

private class InstrumentationArgumentsConverter : IStringConverter<List<String>> {
    override fun convert(argument: String): List<String> = listOf(argument)
}
